pkg/scheduler/initialsizeclass/platform_overriding_size_class_analyzer.go

```diff
@@ -0,0 +1,91 @@
+package initialsizeclass
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	remoteexecution "github.com/bazelbuild/remote-apis/build/bazel/remote/execution/v2"
+	"github.com/buildbarn/bb-storage/pkg/digest"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type platformOverridingSizeClassAnalyzer struct {
+	platformKey            string
+	actionTimeoutExtractor *ActionTimeoutExtractor
+	base                   Analyzer
+}
+
+func NewPlatformOverridingAnalyzer(actionTimeoutExtractor *ActionTimeoutExtractor, platformKey string, base Analyzer) Analyzer {
+	return &platformOverridingSizeClassAnalyzer{
+		platformKey:            platformKey,
+		actionTimeoutExtractor: actionTimeoutExtractor,
+		base:                   base,
+	}
+}
+
+type sizeClassSelector struct {
+	sizeClass string
+	timeout   time.Duration
+}
+
+func (a *platformOverridingSizeClassAnalyzer) Analyze(ctx context.Context, digestFunction digest.Function, action *remoteexecution.Action) (Selector, error) {
+	timeout, err := a.actionTimeoutExtractor.ExtractTimeout(action)
+	if err != nil {
+		return nil, err
+	}
+
+	platform := action.Platform
+	for _, key := range platform.Properties {
+		if key.Name == a.platformKey {
+			return sizeClassSelector{
+				sizeClass: key.Value,
+				timeout:   timeout,
+			}, nil
+		}
+	}
+
+	if a.base == nil {
+		return nil, status.Error(codes.InvalidArgument, "No size class provided and no analyzer configured")
+	}
+
+	return a.base.Analyze(ctx, digestFunction, action)
+}
+
+type emptyLearner struct{}
+
+func (s sizeClassSelector) Select(sizeClasses []uint32) (int, time.Duration, Learner, error) {
+	learner := &emptyLearner{}
+	if s.sizeClass == "max" {
+		return len(sizeClasses) - 1, s.timeout, learner, nil
+	}
+	if s.sizeClass == "min" {
+		return 0, s.timeout, learner, nil
+	}
+	class, err := strconv.Atoi(s.sizeClass)
+	if err != nil {
+		return 0, s.timeout, learner, status.Errorf(codes.InvalidArgument, "Invalid size class '%s'", s.sizeClass)
+	}
+	for idx, val := range sizeClasses {
+		if int(val) == class {
+			return idx, s.timeout, learner, nil
+		}
+	}
+	return 0, s.timeout, learner, status.Errorf(codes.InvalidArgument, "Size class '%s' not available", s.sizeClass)
```

Comment:
I'm not sure that this is solid.

We currently only allow you to use feedback driven size class analysis by predeclaring size classes and providing a maximum size class. That way we can be certain that the system behaves deterministically, even if workers for larger size classes are (temporarily) offline. The client can be sure that if execution fails, it will always run on the largest worker.

The behaviour as implemented will cause hard execution failures, even if workers for a given size class are only temporarily offline. This may also be the case if the scheduler has just restarted and workers are still in the process of reconnecting.